### DISCUSSION TITLE
Should custom `FastApi.openapi` implementations be allowed to be `async`?

### DISCUSSION THREAD

**Discussion by bkis** [fastapi/fastapi#10637] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.openapi.utils import get_openapi

app = FastAPI()

async def custom_openapi():
    settings = await get_settings_from_db()  # hypothetical DB call
    return get_openapi(
        title=settings.get("title", "FastApi OpenAPI"),
        version="2.5.0",
        summary="This is a very custom OpenAPI schema with a title taken from the DB",
        description="Here's a longer description of the custom **OpenAPI** schema",
        routes=app.routes,
    )

app.openapi = custom_openapi
```

### Description

**TL;DR:** Is there anything that speaks against allowing the application's `openapi` method to be `async`?

#### The Problem

The documentation explains how to [override the `openapi` method](https://fastapi.tiangolo.com/how-to/extending-openapi/?h=extend#override-the-method) of the application instance to customize the OpenAPI schema generation.

Now I think that a very prevalent reason to do so is the need to incorporate settings of some kind into the schema. These settings *may* come from something like environment variables that are processed by Pydantic `BaseSettings` where they're read before app initialization. But they may also very well be dynamic application settings that can change at runtime, e.g. when an admin updates the application title via some administration UI.

Such changes have to be stored and persisted somewhere, most probably a database or something like Redis (or both, for that matter). Now accessing such settings often involves `async` calls to e.g. the database (see my code example). But this would – of course – fail with a `TypeError: Object of type coroutine is not JSON serializable`.

#### The Proposal

So to allow for `async` logic to be used in the generation of the OpenAPI schema within `FastAPI.openapi`, the internal call (`return JSONResponse(self.openapi())` in `applications.py`) would be required to handle async `openapi()` implementations.

As the call to `self.openapi()` happens in an async context anyway, the changes needed would be minimal and only introduce one conditional that's run once in the application setup routine (of course this last part largely depends on how the custom `openapi` method is implemented, there might be good reasons to not cache the schema and fetch fresh settings data from the DB).

#### Additional Considerations

This wouldn't even need a change of the documentation. It would simply make FastApi support async implementations of `openapi` in case people need it (as you might have guessed, I am one of those people).

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.10.12

**Comment by bkis** [comment]
In case this is seen as a viable proposal, I have opened a (quite minimalistic) PR offering the implementation: #10638

**Comment by YuriiMotov** [comment]
https://github.com/fastapi/fastapi/pull/10638#issuecomment-2999922298
